falcon_multiqc/commands/sql.py

```diff
@@ -0,0 +1,52 @@
+import click
+import glob
+import os
+from sqlalchemy import create_engine
+from database.config import DATABASE_URI
+from database.process_query import run_multiqc
+
+# example query: 
+
+# this query will query fastqc tool for Total seqence metric - since only the first batch ran fastqc it will only return first batch
+# SELECT s.sample_name, b.path FROM public.raw_data as rd join sample as s on (s.id = rd.sample_id) join batch as b on (s.batch_id = b.id) where rd.qc_tool = 'fastqc' AND (rd.metrics->>'Total Sequences')::numeric::integer >= 316624794;
+
+# this queries two different tools (picard_insertsize and verifybamid) and on of their respective metrics - returns sampleID from first batch and sampleID from 4012 batch 
+# SELECT s.sample_name, b.path FROM public.raw_data as rd join sample as s on (s.id = rd.sample_id) join batch as b on (s.batch_id = b.id) where ((rd.qc_tool = 'picard_insertSize' AND (rd.metrics->>'MEAN_INSERT_SIZE')::numeric::integer <= 390) OR (rd.qc_tool = 'verifybamid' AND (rd.metrics->>'#READS')::numeric::integer >= 4850000));
+
+@click.command()
+@click.option("-d", "--directory", type=click.STRING, required=True, help="where query result will be saved")
+@click.option("-s", "--sql", type=click.INT, required=False, help="Enter 1 to start SQL querying tool") 
+def cli(directory, sql):
+    if sql == 1:
+        """SQL query tool: ensure all queries SELECT for sample_name from sample table AND path from batch table"""
+
+        count = 0 # to ensure new file name generated
+        config = os.path.abspath("./database/falconqc.config") # loading falconqc config file 
+        query_eng = create_engine(DATABASE_URI) # establishing engine for communication 
+        click.echo("Entering SQL querying tool, to exit type 'exit' as the SQL statement")
+        while True:
+            sql = click.prompt("Enter SQL query")
+            if sql == 'exit':
+                break   
+            map_path_sample = {} # Dictionary to store path:sampleID key value pairs 
+            with query_eng.connect() as conn:
+                sample_list = list(conn.execute(sql)) # executes SQL query against database - sample list stores tuples of (path, sampleID)
```

Comment:
I have an idea that could improve performance slightly here. Instead of using list() to convert it all to a list, just use the raw result object.

Then in line 35+ you would go:

                for row in sample_list:
                    try:
                        map_path_sample[row['path']].append(row['sample_name']) 
                    except KeyError:
                        map_path_sample[row['path']] = [row['sample_name']]

Printing out the number of rows is sample_list.rowcount.

Although with this method, once you iterate the rows they are gone. So don't print the rows out before you store them in your dict.

Reply:
Yep done